### ISSUE TITLE
Standard openSesame pipeline (sesame version 1.8.2) does not perform detection masking

### ISSUE THREAD

**Issue by jcm6t** [zwdzwd/sesame#29]
Thank you for supporting the sesame code and pipeline.

Bioconductor version 3.12 (BiocManager 1.30.12),

The study has data from 458 unrelated ppts on EPIC array.

Using standard sesameQC:
sqc <- do.call(rbind, lapply(ssets, function(x) as.data.frame(sesameQC(x))))

we were surprised to find that all 458 had identical fraction of missing beta values.
All samples:
frac_na_cg = 0.1215676        
frac_nondt = 0

In 458 arrays we would have expected at least some non-detected probe variability. The results suggested to us that the pipeline was flagging probe-level issues (qualityMask) but not individual probe-to-probe detection.

snippet of code:
ssets.nb.corr <- lapply(ssets, function(sset) { dyeBiasCorrTypeINorm(noob(sset)) })
ssets.mask <- lapply(ssets.nb.corr, detectionMask)

33 seems to be a poorer quality array, so use this as an example:
ssets.mask[[33]]

SigSet EPIC
 - @IG probes: 49946 - 221.6352 3210.656 867.3447 9250.656 558.0544 3941.656 ...
 - @IR probes: 92212 - 185.7184 2411.447 186.335 9258.156 684.9406 178.4422 ...
 - @II probes: 723760 - 4927.156 1231.822 2079.383 278.3062 190.42 203.4235 ...
 - @oobG probes: 92212 - 188.8977 203.9813 191.1679 292.1194 212.4178 178.6639 ...
 - @oobR probes: 49946 - 219.0378 195.3325 188.6529 1008.59 204.4254 194.9974 ...
 - @ctl probes: 635 ...
 - @pval: 0 -  ...

ssets[[33]]@extra$mask

NULL

in detectionMask ( since we running with default options and pval.method = NULL :)

 if (is.null(pval.method)) {
        pv <- pval(sset)
    }

Looking in the code it seems that there is a standardGeneric defined for pval but there don't appear to be any methods.
In other words, pval() is basically doing nothing.

pv<-sesame:::pval(ssets[[33]])
summary(pv)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.

It would be helpful if you could confirm a workaround.

For other users - please note that if you only look at a single sset at the end of the pipeline, you wouldn't be able to tell that anything was wrong without digging further since probe-level masking seems to work.

Thanks

**Comment by ir** [comment]
the joys of having a 2 letter github username

**Comment by zwdzwd** [comment]
Thanks for reporting this. But did you use openSesame to generate the betas? In readIDATpair function, the default pvalue is set using pOOBAH. So the pval slot in  sset shouldn't be 0. I just tested using 1.8.2 it seems I do get sample-specific masking

```
> betas = openSesame(".")
> colSums(is.na(betas))
GSM2178224 GSM2178226
    182277     235427  
```

`pval(sset)` is simply a getter function of sset@pval, that's the default pval, alternatives are put to `sset@extra[['pvals']]`

@ir sorry, it was supposed to be "IG". Speaking of that, I am curious, what platform are you on? the print result seems to be very different from my side.

**Comment by zwdzwd** [comment]
For example, when I load readIDATpair, I actually get non-empty pval slot by default. 

```
> readIDATpair("GSM2178224")
SigSet EPIC
 - @IG probes: 49946 - 519 7650 510 14210 944 4732 ...
 - @IR probes: 92212 - 1076 8176 3560 17882 1542 1212 ...
 - @II probes: 723760 - 9263 3164 4913 1264 1154 1061 ...
 - @oobG probes: 92212 - 746 301 459 751 527 496 ...
 - @oobR probes: 49946 - 1700 764 881 2915 998 792 ...
 - @ctl probes: 635 ...
 - @pval: 865918 - 0.001101189 0.006156649 0.0484223 ...
 ```

**Comment by jcm6t** [comment]
zwdzwd,
Thanks for your quick response.

No, the ssets object was created from 
ssets <- RGChannelSetToSigSets(RGSet)

This was because we were comparing minifi pipeline with sesame. So it appears that there is some additional hidden processing in the readIDATpair function that is not described, ie 
https://github.com/zwdzwd/sesame/blob/master/man/readIDATpair.Rd

If this is true, then I would strongly suggest that you limit the readIDATpair function to only reading data and creating the S4 object structure. By all means, wrapper the pOOBAH step into openSesame but expose it as a discrete step. Alternatively, could you combine it into the qualityMask step which is where we assumed pOOBAH was happening.

The next question - we can probably guess now but it would be good for you to confirm - exactly what p-value processing is happening in:
readIDATpair
vs
detectionMask 
vs 
qualityMask
?

Just a final observation. For our 458 samples:
RGSET RData file:  4.0 GB (this contains the RGSet and the Genome Methyl Set object annotated against EPIC ilm10b5)
converted to SSets:
SSets RData file:  8.4 GB
To be clear we are only saving the ssets object in this file, and neither the RGSet nor the GMSet object is included.  So the data structure looks a little bloated compared to RGSet which, whether you love it or hate it, is more of a de facto standard object in Bioconductor with a longer history.

Thanks very much for your help.

**Comment by zwdzwd** [comment]
re: readIDATpair only does data reading. That's a good suggestion. That was actually one of the changes in 1.9+. Currently there is still a poobah call at the end of readIDATpair (the hidden processing you mentioned). This gives a default p-value and prevents accidental execution of detectionMask without detection calling.

detectionMask actually doesn't calculate p-value. It uses p-value previously calculated (using pval.method argument) to create a masking. The masking is then optionally used in getBetas to produce beta values. This arrangement is to allow multiple p-value calculation. Both detectionMask and qualityMask alters masking. But the masking is not executed to the probe signal until getBetas. Other words, the masking can be altered and reset, as long as we don't call getBetas. That would allow flexible creation of the masking.

Re the increase in the sset rdata: it is because of the @extra column that stores the pvalues and the masking. In the recent versions, one has the option of removing the interim p-values and masking. That should reduce the size to a number comparable to RGset. There is also fileSet option which compress the data further if storage is a concern. Most of the sesame workflow is independent of RGSet and GMSet. We specifically added sesamize for minfi-interaction because these RGSet/GMSet formats are more popular and used in the field for a longer history, like you said. 

Hope this clarifies the logic a bit.

**Comment by jcm6t** [comment]
That is helpful. Probably be good to add more detail to the User Guide when you have the chance but not urgent for this issue.

So it looks like there is a bug in RGChannelSetToSigSets(). It there a quick fix possible to 1.8.2 or is the workaround to rerun the pipeline starting with readIDATpair() ? 

Once you confirm, we can close this.

**Comment by zwdzwd** [comment]
I have added the following call to RGChannelSetToSigSets().
https://github.com/zwdzwd/sesame/blob/490d0bf6525d1cca950d73f0efafec771955ccab/R/sesamize.R#L244
It will hopefully address this bug. This change has been pushed to both release and devel. 

Thanks for the suggestion. I will add more detail to the User Guide.

**Comment by jcm6t** [comment]
Testing Release_3_12. (1.8.7)

So It appears that the p-values have been added to the object and the processing is working. 

But after calling getBetas() with the default mask=T we still do not see missing individual beta values.
When you mask the sporadic non-detected probes what should getBetas() set the beta value to ? 
We assumed it would be NA.

Thanks.

**Comment by jcm6t** [comment]
Explaining the last comment in more detail with code.

sesame version 1.8.7

1.  sesameQC appears to be working now. 

`ssets <- RGChannelSetToSigSets(RGSet)`
`sqc <- do.call(rbind, lapply(ssets, function(x) as.data.frame(sesameQC(x))))`

`summary(sqc$frac_nondt)`
`#   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.`
`#0.00718 0.02216 0.02976 0.03795 0.04395 0.30615`

But note that frac_na_cg, which we also expected to vary, displays all 0s.
`summary(sheetqc$frac_na_cg)`
`# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.`
`#      0       0       0       0       0       0`

2.  Results for our modified openSesame() pipeline are still a bit uncertain. Perhaps we don't understand the functions properly.
It appears that qualityMask() is also changing data structures under the covers that we are bypassing because of our custom probe drop step, and this is probably why our results do not make sense. We'd like more control over the pipeline ensuring that stepwise functions do not have hidden dependencies that break custom implementations. Again, openSesame() provides that option for labs that just want to run the default pipeline.

`# noob + dyebias`
`ssets.nb.corr <- lapply(ssets, function(sset) { dyeBiasCorrTypeINorm(noob(sset)) })`
`ssets.mask <- lapply(ssets.nb.corr, detectionMask)`
`# DO NOT apply generic qualityMask().   Read in the data file from Zhou website specific for the EUR population and remove those probes manually - after getBetas`

`#read in the Zhou 2017 European specific black list`
`Zhou.EUR<-read.delim("../annot/EPIC.hg38.EUR_filter_probes.tsv",head=T,sep="\t")`
`EUR.dropprobes<-as.vector(Zhou.EUR$probeID)`

`betas <- lapply(ssets.mask, getBetas)`
`nprobe<-length(betas[[1]])`
`probenames<-names(betas[[1]])`
`betamtx<-matrix(unlist(betas),byrow=T,ncol=nprobe)`
`dim(betamtx)`
`#[1]    458 865918`
`rownames(betamtx)<-names(betas)`
`colnames(betamtx)<-probenames`
`betamtx <- betamtx[,!colnames(betamtx) %in% EUR.dropprobes]`
`dim(betamtx)`
'#[1]    458 769995'

`# now summarize missing rates in remaining probes`
`betamiss <- is.na(betamtx)`
`nNA <- rowSums(betamiss)`
`fracNA<-nNA/nprobe`
`summary(fracNa)`
`# Min. 1st Qu.  Median    Mean 3rd Qu.    Max.`
`#      0       0       0       0       0       0`

Hence all remaining probes have 0 missing rate. We would have expected residual sporadic missing betas (NAs) in the data set even in the retained probes.

Hope this helps,
Thanks.

**Comment by zwdzwd** [comment]
I am trying using the following test data which seem to generate non-zero result?
> ssets <- RGChannelSetToSigSets(FlowSorted.Blood.450k[,1:4])
> sqc <- do.call(rbind, lapply(ssets, function(x) as.data.frame(sesameQC(x))))
> summary(sqc$frac_na_cg)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.1445  0.1493  0.1568  0.1559  0.1633  0.1655

To have more control, I would recommend simply use the following equivalent of openSesame
```r
betas <- do.call(cbind, lapply(searchIDATprefixes(idat_dir), function(pfx) {
    pfx %>%
        readIDATpair %>%
        pOOBAH %>%
        noob %>%
        dyeBiasCorrTypeINorm %>%
        qualityMask %>%
        detectionMask %>%
        getBetas
}))
```

All what `qualityMask` does is that it changes the `sset@extra$mask` object by adding the probes with non-significant pvalue 
to the mask. The related code is here.
https://github.com/zwdzwd/sesame/blob/312ac2a15d4e25c4c8667c25c8d1241505c10f6d/R/mask.R#L125

I am not sure I fully understand your error. but looks like if you do have detectionMask and proper p-value calculation, you should be able to get masking?

**Comment by jcm6t** [comment]
Here are the results for the default pipeline using function calls, except that we are starting from RGSet. We've not going to write code to scan all of the raw dirs to match all 458 pairs to read each pair individually.

```
ssets <- RGChannelSetToSigSets(RGSet)
# noob + dyebias
ssets.nb.corr <- lapply(ssets, function(sset) { dyeBiasCorrTypeINorm(noob(sset)) })
ssets.mask <- lapply(ssets.nb.corr, qualityMask)
ssets.mask <- lapply(ssets.mask, detectionMask)

#print("getBetas with mask defaulted to T")
betas <- lapply(ssets.mask, getBetas)
betas[[1]][1:10]

> betas[[1]][1:10]
cg00000029 cg00000103 cg00000109 cg00000155 cg00000158 cg00000165 cg00000221 cg00000236 cg00000289 cg00000292
 0.7105836  0.8868143  0.8774039  0.9381188  0.9413551  0.1303788  0.8474027  0.8242698  0.6866013  0.9046317

nprobe<-length(betas[[1]])
nprobe
[1] 865918
probenames<-names(betas[[1]])

betamtx<-matrix(unlist(betas),byrow=T,ncol=nprobe)
dim(betamtx)
[1]    458 865918

rownames(betamtx)<-names(betas)
colnames(betamtx)<-probenames

> betamtx[1:3,1:8]
           cg00000029 cg00000103 cg00000109 cg00000155 cg00000158 cg00000165 cg00000221 cg00000236
40006049    0.7105836  0.8868143  0.8774039  0.9381188  0.9413551  0.1303788  0.8474027 0.9046317
40006061    0.6569820  0.9183372  0.9256316  0.9485566  0.9565295  0.1591430  0.9016238  0.8327268  
40006073    0.4789715  0.8746143  0.8838724  0.9261565  0.9311672  0.1254480  0.8663420  0.8162297 

betamiss <- is.na(betamtx)
nNA <- rowSums(betamiss)
summary(nNA/nprobe)
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
      0       0       0       0       0       0

summary(betamtx[1,])
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
 0.0103  0.2438  0.8041  0.6184  0.9099  0.9890

```
At this point I am at a loss other than we will write our code around qualityMask and detectionMask. Dropping probes is a trivial step but it would be nice to make sure we detecting probes per the latest recommendations. We'll just have to expose the p-value processing.

Unless you have other suggestions?
If not, we'll just close this half-resolved.

FYI,
After the pOOBAH call, the converted ssets RData file expanded from 8 to 12GB.

Thanks,

**Comment by zwdzwd** [comment]
Sorry for the frustration. The example I showed you is also based on RGSet. I am trying the same command as yours and got the following.

```{r}
library(FlowSorted.Blood.450k)
RGSet <- FlowSorted.Blood.450k[,1:4]

## > RGSet
## class: RGChannelSet
## dim: 622399 4
## metadata(0):
## assays(2): Green Red
## rownames(622399): 10600313 10600322 ... 74810490 74810492
## rowData names(0):
## colnames(4): WB_105 WB_218 WB_261 PBMC_105
## colData names(8): Sample_Name Slide ... CellType Sex
## Annotation
##   array: IlluminaHumanMethylation450k
##   annotation: ilmn12.hg19

ssets <- RGChannelSetToSigSets(RGSet)
# noob + dyebias
ssets.nb.corr <- lapply(ssets, function(sset) { dyeBiasCorrTypeINorm(noob(sset)) })
ssets.mask <- lapply(ssets.nb.corr, qualityMask)
ssets.mask <- lapply(ssets.mask, detectionMask)

#print("getBetas with mask defaulted to T")
betas <- lapply(ssets.mask, getBetas)
betas[[1]][1:10]

> betas[[1]][1:10]
cg00000029 cg00000108 cg00000109 cg00000165 cg00000236 cg00000289 cg00000292
0.61836373 0.96733429 0.86443541         NA 0.90052965 0.69864631 0.89254265
cg00000321 cg00000363 cg00000622
0.29782104 0.38974849 0.01537851

nprobe<-length(betas[[1]])
nprobe
[1] 485577
probenames<-names(betas[[1]])

betamtx<-matrix(unlist(betas),byrow=T,ncol=nprobe)
dim(betamtx)
[1]      4 485577

rownames(betamtx)<-names(betas)
colnames(betamtx)<-probenames

> betamtx[1:3,1:8]
       cg00000029 cg00000108 cg00000109 cg00000165 cg00000236 cg00000289
WB_105  0.6183637  0.9673343  0.8644354         NA  0.9005296  0.6986463
WB_218  0.5394411  0.9567713  0.8665002  0.1798287  0.8317031         NA
WB_261  0.5641759  0.9660559  0.8979344  0.1782666  0.8544802         NA
       cg00000292 cg00000321
WB_105  0.8925426  0.2978210
WB_218  0.9039978  0.2962243
WB_261  0.8781822  0.2982283

betamiss <- is.na(betamtx)
nNA <- rowSums(betamiss)
summary(nNA/nprobe)
##   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.
## 0.1456  0.1507  0.1583  0.1574  0.1651  0.1674
```

My sessionInfo()
 [1] sesame_1.8.7                 sesameData_1.8.8

Another suggestion: what if we explicitly add a pOOBAH call like
```r
ssets.nb.corr <- lapply(ssets, function(sset) { dyeBiasCorrTypeINorm(noob(pOOBAH(sset))) })
```
Does that give us some NA?

**Comment by jcm6t** [comment]
Before we saw your email we re-ran previous but outputting summary of one sset sset[[1]], at each step.
The p-values are set, but the mask handling appears a bit less stable.

More generally, you are testing using 450K. Is it possible this is an EPIC issue even through the object structures are the same ?
Generally, there ought to be an EPIC data set to test changes also.

```
.. as previously from RGSet...

# noob + dyebias
ssets.nb.corr <- lapply(ssets, function(sset) { dyeBiasCorrTypeINorm(noob(sset)) })

print("Pre qualityMask, Post-noob, dye bias")
[1] "Pre qualityMask, Post-noob, dye bias"


ssets.nb.corr[[1]]
SigSet EPIC
 - @IG probes: 49946 - 257.9209 4041.509 1360.651 10698.51 755.1263 3729.009 ...
 - @IR probes: 92212 - 595.7885 5323.509 462.2873 10310.01 401.0166 183.4407 ...
 - @II probes: 723760 - 6651.474 2452.495 4508.009 324.1648 226.9361 242.734 ...
 - @oobG probes: 92212 - 299.8587 289.1512 245.7118 266.2175 258.3285 190.8832 ...
 - @oobR probes: 49946 - 262.5783 209.3746 212.6084 824.922 260.2337 219.7459 ...
 - @ctl probes: 635 ...
 - @pval: 865918 - 0.000480519 0.004775157 0.2302587 ...

names(sesame:::extra(ssets.nb.corr[[1]]))
[1] "pvals"

ssets.nb.corr[[1]]@extra$mask
NULL

ssets.mask <- lapply(ssets.nb.corr, qualityMask)
print("Post qualityMask")
[1] "Post qualityMask"

ssets.mask[[1]]
SigSet EPIC
 - @IG probes: 49946 - 257.9209 4041.509 1360.651 10698.51 755.1263 3729.009 ...
 - @IR probes: 92212 - 595.7885 5323.509 462.2873 10310.01 401.0166 183.4407 ...
 - @II probes: 723760 - 6651.474 2452.495 4508.009 324.1648 226.9361 242.734 ...
 - @oobG probes: 92212 - 299.8587 289.1512 245.7118 266.2175 258.3285 190.8832 ...
 - @oobR probes: 49946 - 262.5783 209.3746 212.6084 824.922 260.2337 219.7459 ...
 - @ctl probes: 635 ...
 - @pval: 865918 - 0.000480519 0.004775157 0.2302587 ...

names(sesame:::extra(ssets.mask[[1]]))
[1] "pvals"

ssets.mask[[1]]@extra$mask
NULL

summary(ssets.mask[[33]]@extra$mask)
Length  Class   Mode
     0   NULL   NULL

ssets.mask <- lapply(ssets.mask, detectionMask)

print("Post detectionMask")
[1] "Post detectionMask"

ssets.mask[[1]]
SigSet EPIC
 - @IG probes: 49946 - 257.9209 4041.509 1360.651 10698.51 755.1263 3729.009 ...
 - @IR probes: 92212 - 595.7885 5323.509 462.2873 10310.01 401.0166 183.4407 ...
 - @II probes: 723760 - 6651.474 2452.495 4508.009 324.1648 226.9361 242.734 ...
 - @oobG probes: 92212 - 299.8587 289.1512 245.7118 266.2175 258.3285 190.8832 ...
 - @oobR probes: 49946 - 262.5783 209.3746 212.6084 824.922 260.2337 219.7459 ...
 - @ctl probes: 635 ...
 - @pval: 865918 - 0.000480519 0.004775157 0.2302587 ...

names(sesame:::extra(ssets.mask[[1]]))
[1] "pvals" "mask"

ssets.mask[[1]]@extra$mask
logical(0)

summary(ssets.mask[[33]]@extra$mask)
   Mode
logical

print("getBetas with mask defaulted to T")
[1] "getBetas with mask defaulted to T"

betas <- lapply(ssets.mask, getBetas)
betas[[1]][1:10]
cg00000029 cg00000103 cg00000109 cg00000155 cg00000158 cg00000165 cg00000221 cg00000236 cg00000289 cg00000292
 0.7105836  0.8868143  0.8774039  0.9381188  0.9413551  0.1303788  0.8474027  0.8242698  0.6866013  0.9046317

...etc.. as previously
```

**Comment by zwdzwd** [comment]
I think I have eventually caught the culprit! I missed `sset <-`  in `sset <- resetMask(sset);` in `detectionMask`.
The fix has been uploaded to 1.8.10. It should've been fixed now. Many thanks for the feedback!

Temporary link
https://www.dropbox.com/s/nskhg2qztz4wnwm/sesame_1.8.10.tar.gz?dl=0

**Comment by jcm6t** [comment]
This seems to be working now, so I am going to close this. Thanks for your help.

**Comment by saeed-svu** [comment]
Hi, 
it is my first comment on github!

I started using SeSAME recently for my thesis. When checking the number of masked probes per sample for DIFFERENT experiments, i am surprised to find the same number of probes is masked among different experiments(arrays)!

Specifically 64144 probes is masked when the platform is HM450K and 105454 probes when the platform is EPIC v1. So no matter what experiment , The samples will always will have this amount of masked probes. To confirm i extracted a csv table summarizing the masked probes in every standalone step for every experiment (before and after quality mask, masked probes from pOOBAH alone, masked probes from noob and dyeBias (zero but i include it any way).

Can anyone explain? is it normal?

by the way, pOOBAH masking is sample speicific, as i noticed different number of masked probes in every sample in every array. But the identical numbers was observed in quality mask step. 

## Code and steps i followed

s <- readIDATpair(file)

# Check if SeSAME Identifies the platform:
platform <- sdfPlatform(s)

# Step 1: Check masked probes after reading idats:
initial_mask <- sum(s$mask)       # 0 masked probes

# Step 2: Apply Quality Mask
s1 <- qualityMask(s)
quality_mask <- sum(s1$mask)       #105454 masked probes when EPIC or 64144 when 450K

# Step 3: Apply dyeBiasNL (mask not modified)
s2 <- dyeBiasNL(s)      
dyeBiasNL_mask <- sum(s2$mask)                   # zero masked probes

# Step 4: Apply pOOBAH over the original data
s3 <- pOOBAH(s)
pOOBAH_mask <- sum(s3$mask)     # different number for each sample (approximately 4K - 14K probes)

# Step 5: Apply noob
s4 <- noob(s)
noob_mask <- sum(s4$mask)    # 0 masked probes



Thanks in advance.
